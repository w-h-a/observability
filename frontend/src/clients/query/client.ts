export interface IClient {
	get<T = any>(url: string): Promise<{ data: T }>;
}
