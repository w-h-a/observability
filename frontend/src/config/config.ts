export enum EnvVar {
	ENVIRONMENT = "ENVIRONMENT",
	BASE_QUERY_URL = "BASE_QUERY_URL",
}

export class Config {
	private static Instance: Config;
	private configData: Record<EnvVar, string>;

	private constructor() {
		const envVars = process.env;

		this.configData = {
			[EnvVar.ENVIRONMENT]: "dev",
			[EnvVar.BASE_QUERY_URL]: "http://localhost:4000/api/v1",
		};

		if (envVars["NODE_ENV"]) {
			this.configData.ENVIRONMENT = envVars["NODE_ENV"];
		}

		if (envVars["BASE_QUERY_URL"]) {
			this.configData.BASE_QUERY_URL = envVars["BASE_QUERY_URL"];
		}
	}

	public static GetInstance(): Config {
		if (!Config.Instance) {
			Config.Instance = new Config();
		}

		return Config.Instance;
	}

	public get(key: EnvVar): string {
		return this.configData[key];
	}
}
