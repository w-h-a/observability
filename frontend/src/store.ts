import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { thunk } from "redux-thunk";

export const reducers = combineReducers({});

export const store = createStore(reducers, compose(applyMiddleware(thunk)));
