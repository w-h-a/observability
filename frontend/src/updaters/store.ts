import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { ServicesUpdater } from "./services/servicesTable";
import { TimeUpdater } from "./time/time";
import { ServiceUpdater } from "./service/service";

const reducer = combineReducers({
	maxMinTime: TimeUpdater.MaxMinTimeReducer,
	services: ServicesUpdater.ServicesReducer,
	endpoints: ServiceUpdater.EndpointsReducer,
});

export const store = configureStore({
	reducer: reducer,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
