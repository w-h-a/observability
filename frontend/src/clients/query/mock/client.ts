import { IClient } from "../client";

export class Client implements IClient {
	async get<T = any>(url: string): Promise<{ data: T }> {
		return {
			data: [] as T,
		};
	}
}
