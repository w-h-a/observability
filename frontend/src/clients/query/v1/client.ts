import axios, { AxiosInstance } from "axios";
import { IClient } from "../client";

export class Client implements IClient {
	private client: AxiosInstance = axios.create();

	async get<T = any>(url: string): Promise<{ data: T }> {
		const result = await this.client.get<T>(url);
		return result;
	}
}
