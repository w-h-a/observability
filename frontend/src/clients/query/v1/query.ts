import { Config } from "../../../config/config";
import { IClient } from "../client";

export class Query {
	static async GetServices<T = any>(
		client: IClient,
		start: number,
		end: number,
	): Promise<{ data: T }> {
		try {
			const path = `/services`;

			const query = `?start=${start}&end=${end}`;

			return await client.get<T>(
				`${Config.GetInstance().get("baseUrl")}${path}${query}`,
			);
		} catch (err: unknown) {
			console.log(`query client failed to retrieve services: ${err}`);
			throw err;
		}
	}

	static async GetEndpoints<T = any>(
		client: IClient,
		start: number,
		end: number,
		serviceName: string,
	): Promise<{ data: T }> {
		try {
			const path = `/service/endpoints`;

			const query = `?start=${start}&end=${end}&service=${serviceName}`;

			return await client.get<T>(
				`${Config.GetInstance().get("baseUrl")}${path}${query}`,
			);
		} catch (err: unknown) {
			console.log(`query client failed to retrieve endpoints: ${err}`);
			throw err;
		}
	}
}
