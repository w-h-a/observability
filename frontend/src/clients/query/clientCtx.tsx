import { createContext } from "react";
import { Client } from "./v1/client";
import { IClient } from "./client";

export const ClientContext = createContext<{ queryClient: IClient }>({
	queryClient: new Client(),
});
