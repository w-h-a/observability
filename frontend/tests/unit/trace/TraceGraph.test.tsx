import { Config } from "../../../src/config/config";

Config.GetInstance();

import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { render, RenderResult } from "@testing-library/react";
import { TestCase } from "../testcase";
import { TraceGraph } from "../../../src/views/Trace/TraceGraph";
import { store } from "../../../src/updaters/store";
import { IClient } from "../../../src/clients/query/client";
import { Client } from "../../../src/clients/query/mock/client";
import { ClientContext } from "../../../src/clients/query/clientCtx";

const traceId = "8fb945b2df69da477a98886d79c9a26d";

vi.mock("react-router-dom", async () => {
	const mod = await vi.importActual("react-router-dom");
	return {
		...mod,
		useParams: () => ({
			id: traceId,
		}),
	};
});

describe("TraceGraph", () => {
	const mockQueryClient: IClient = new Client();

	const testCases: TestCase[] = [
		{
			When:
				"when: the user goes to the spans page and calls to the backend are a success",
			GetStub: function <T>(url: string): Promise<{ data: T }> {
				return new Promise((resolve) => {
					resolve({
						data: [
							{
								columns: [
									"Time",
									"SpanId",
									"ParentSpanId",
									"TraceId",
									"ServiceName",
									"Name",
									"Kind",
									"StatusCode",
									"Duration",
									"Tags",
								],
								events: [
									[
										1737306333967,
										"100ab2aea075c0d5",
										"bfcf37f4c5b5568c",
										"8fb945b2df69da477a98886d79c9a26d",
										"route",
										"/route",
										"Server",
										"Ok",
										"44705750",
										[
											["http.target", "/route"],
											["http.status_code", "200"],
										],
									],
									[
										1737306333966,
										"bfcf37f4c5b5568c",
										"",
										"8fb945b2df69da477a98886d79c9a26d",
										"frontend",
										"HTTP GET",
										"Client",
										"Ok",
										"45147416",
										[
											["http.status_code", "200"],
											["http.method", "GET"],
											[
												"http.url",
												"http://0.0.0.0:8083/route?dropoff=728%2C326&pickup=947%2C38",
											],
										],
									],
								],
							},
						] as T,
					});
				});
			},
			Then: "then: we render the spans table",
			ClientCalledTimes: 1,
			ClientCalledWith: [
				`http://localhost:4000/api/v1/spans/trace?traceId=${traceId}`,
			],
		},
	];

	testCases.forEach((testCase) => {
		describe(testCase.When, () => {
			let view: RenderResult;

			beforeEach(async () => {
				mockQueryClient.get = vi.fn().mockImplementation(testCase.GetStub);

				view = render(
					<Provider store={store}>
						<BrowserRouter>
							<ClientContext.Provider value={{ queryClient: mockQueryClient }}>
								<TraceGraph />
							</ClientContext.Provider>
						</BrowserRouter>
					</Provider>,
				);
			});

			afterEach(() => {
				vi.clearAllMocks();
			});

			test(testCase.Then, async () => {
				expect(mockQueryClient.get).toHaveBeenCalledTimes(
					testCase.ClientCalledTimes,
				);

				for (let count = 1; count <= testCase.ClientCalledTimes; count++) {
					expect(mockQueryClient.get).toHaveBeenCalledWith(
						testCase.ClientCalledWith[count - 1],
					);
				}

				expect(view.container).toMatchSnapshot();
			});
		});
	});
});
