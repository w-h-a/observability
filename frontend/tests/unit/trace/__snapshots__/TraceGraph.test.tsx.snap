// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TraceGraph > when: the user goes to the spans page and calls to the backend are a success > then: we render the spans table 1`] = `
<div>
  <div
    class="ant-row css-dev-only-do-not-override-7ny38l"
    style="margin-left: -16px; margin-right: -16px;"
  >
    <div
      class="ant-col ant-col-sm-24 ant-col-md-24 css-dev-only-do-not-override-7ny38l"
      style="padding-left: 16px; padding-right: 16px;"
    >
      <div
        class="ant-space css-dev-only-do-not-override-7ny38l ant-space-vertical ant-space-gap-row-middle ant-space-gap-col-middle"
        style="width: 100%;"
      >
        <div
          class="ant-space-item"
        >
          <div
            class="ant-card ant-card-bordered css-dev-only-do-not-override-7ny38l"
            style="height: 500px;"
          >
            <div
              class="ant-card-body"
              style="padding: 80px;"
            >
              <div
                style="display: flex; justify-content: center; flex-direction: column; align-items: center;"
              >
                <div
                  style="text-align: center; margin-top: 40px;"
                >
                  Trace Graph For Trace 
                  8fb945b2df69da477a98886d79c9a26d
                   
                </div>
                <div
                  id="graph"
                  style="font-size: 12px; margin-top: 80px;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
